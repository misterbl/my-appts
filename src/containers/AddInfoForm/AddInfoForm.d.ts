import { IAppState } from '../../types/state';
import { RouteComponentProps, StaticContext } from 'react-router';
import { InjectedIntlProps } from 'react-intl';
import { ActionCreatorsMapObject } from 'redux';
import { FormikProps } from 'formik';

export interface IAddInfoFormMapStateToProps {
  user: IAppState['api']['userData'];
}

export interface IAddInfoFormData {
  profileTitle?: string;
  profileDescription?: string;
}

export interface IAddInfoFormDispatchToProps {
  apiThunk: ActionCreatorsMapObject;
}

export interface IAddInfoFormComponent {
  submitButton: HTMLButtonElement;
}

export type TAddInfoForm = FormikProps<IAddInfoFormData> &
  IAddInfoFormComponent &
  IAddInfoFormDispatchToProps &
  IAddInfoFormMapStateToProps &
  InjectedIntlProps &
  RouteComponentProps<any, StaticContext>;
