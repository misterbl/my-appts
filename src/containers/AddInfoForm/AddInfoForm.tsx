import * as React from 'react';
import { Formik, Form, FormikProps } from 'formik';

import { bindActionCreators, Dispatch, AnyAction } from 'redux';
import { connect } from 'react-redux';
import { injectIntl, FormattedMessage } from 'react-intl';
import * as Yup from 'yup';
import { withRouter } from 'react-router-dom';
import {
  TAddInfoForm,
  IAddInfoFormDispatchToProps,
  IAddInfoFormMapStateToProps,
} from './AddInfoForm.d';
import * as apiThunk from '../../actions/thunks/apiThunk';
import { IAppState } from '../../types/state';
import { getUserData } from '../../selectors/apiSelectors';
import { QUERIES } from '../../consts';
import FormikInput from '../../components/FormikInput/FormikInput';
import labelColor from '../../utils/labelColor';

export const AddInfoForm = (props: TAddInfoForm) => {
  const {
    values,
    setFieldValue,
    errors,
    intl: { formatMessage },
    submitButton,
  } = props;
  return (
    <Form className="profile-form mh4 mt4 flex flex-column ph5-m ph7-l">
      <FormikInput
        {...props}
        values={values.profileTitle}
        errors={errors.profileTitle}
        setFieldValue={setFieldValue}
        name="profileTitle"
      />

      <label
        className={`${labelColor(values.profileDescription)} f6`}
        htmlFor="profileDescription"
      >
        <FormattedMessage id="general|placeholder|profileDescription" />
      </label>
      <textarea
        className="ba b--light-silver h5 resize-none"
        value={values.profileDescription}
        name="profileDescription"
        onChange={event =>
          setFieldValue('profileDescription', event.target.value)
        }
        placeholder={formatMessage({
          id: 'general|placeholder|profileDescription',
        })}
      />

      {submitButton}
    </Form>
  );
};
export class AddInfo extends React.Component<TAddInfoForm> {
  onSubmit = async (event: any) => {
    const {
      profileTitle: eventProfileTitle,
      profileDescription: eventProfileDescription,
    } = event;
    const {
      // @ts-ignore
      user: { _id },
      user,
    } = this.props;
    const profileTitle =
      event.profileTitle === 'undefined'
        ? user && user.profileTitle
        : eventProfileTitle;

    const profileDescription =
      event.profileDescription === 'undefined'
        ? user && user.profileDescription
        : eventProfileDescription;
    await this.props.apiThunk.updateUser(
      QUERIES({ _id, profileTitle, profileDescription }).UPDATE_AD_INFO,
    );
  };

  render() {
    const {
      intl: { formatMessage },
    } = this.props;

    const { user } = this.props;

    return (
      <Formik
        initialValues={{
          profileTitle: user ? user.profileTitle : '',
          profileDescription: user ? user.profileDescription : '',
        }}
        validationSchema={Yup.object().shape({
          profileTitle: Yup.string().required(
            formatMessage({
              id: 'content|adinfoform|adTitleRequired',
            }),
          ),
          profileDescription: Yup.string().required(
            formatMessage({
              id: 'content|adinfoform|adDescriptionRequired',
            }),
          ),
        })}
        onSubmit={this.onSubmit}
        render={
          /* istanbul ignore next */ formikProps => (
            <AddInfoForm {...formikProps} {...this.props} />
          )
        }
      />
    );
  }
}

export const mapStateToProps = (
  state: IAppState,
): IAddInfoFormMapStateToProps => ({
  user: getUserData(state),
});

export const mapDispatchToProps = (
  dispatch: Dispatch<AnyAction>,
): IAddInfoFormDispatchToProps => ({
  apiThunk: bindActionCreators(apiThunk, dispatch),
});

const injectIntlAddInfoForm = injectIntl(AddInfo);
export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(injectIntlAddInfoForm),
);
