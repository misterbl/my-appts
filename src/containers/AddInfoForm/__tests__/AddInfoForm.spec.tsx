import * as React from 'react';
import { shallow } from 'enzyme';
import formikRenderMock from '../../../testMocks/formikRender.mock';
import historyMock from '../../../testMocks/history.mock';
import { AddInfo } from '../AddInfoForm';
import IntlMock from '../../../testMocks/intl.mock';
import generateAppState from '../../../testMocks/appState.mock';

describe('AddInfoForm', () => {
  const props = {
    values: {
      profileTitle: 'profileTitleTest',
      profileDescription: 'profileDescriptionTest',
    },
    ...formikRenderMock,
    initialValues: {
      profileTitle: '',
      profileDescription: '',
    },
    user: generateAppState().api.userData,
    apiThunk: { updateUser: jest.fn() },
    intl: IntlMock,
    history: historyMock,
    location: historyMock.location,
    match: {
      params: {},
      isExact: true,
      path: '',
      url: '',
    },
    onSubmit: jest.fn(),
    faceBookAddInfoForm: jest.fn(),
  };
  // @ts-ignore
  const wrapper = shallow(<AddInfo {...props} />);
  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a Formik element', () => {
    expect(wrapper.find('Formik').length).toBe(1);
  });

  it('should call updateUser on form submit', () => {
    const {
      values: { profileTitle, profileDescription },
      apiThunk: { updateUser },
    } = props;
    const formik = wrapper.find('Formik');
    formik.simulate('submit', { profileTitle, profileDescription });
    expect(updateUser).toHaveBeenCalled();
  });
  //   describe('Errors', () => {
  //     // @ts-ignore
  //     const addInfoFormForm = shallow(<AddInfoForm {...props} />);
  //     it('it shows an error message for profileTitle when there is an error', () => {
  //       addInfoFormForm.setProps({
  //         errors: { profileTitle: 'profileTitleError' },
  //         touched: { profileTitle: true },
  //       });
  //       const errroMessage = addInfoFormForm.find('ErrorMessage');
  //       expect(errroMessage.length).toBe(1);
  //     });
  //     it('it shows an error message for profileDescription when there is an error', () => {
  //       addInfoFormForm.setProps({
  //         errors: { profileDescription: 'profileDescriptionError' },
  //         touched: { profileDescription: true },
  //       });
  //       const errroMessage = addInfoFormForm.find('ErrorMessage');
  //       expect(errroMessage.length).toBe(1);
  //     });
  //   });
});
