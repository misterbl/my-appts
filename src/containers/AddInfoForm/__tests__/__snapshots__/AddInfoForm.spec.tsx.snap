// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddInfoForm matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddInfo
    apiThunk={
        Object {
            "updateUser": [Function],
          }
    }
    dirty={false}
    errors={Object {}}
    faceBookAddInfoForm={[Function]}
    handleBlur={[Function]}
    handleChange={[Function]}
    handleReset={[Function]}
    handleSubmit={[Function]}
    history={
        Object {
            "action": "PUSH",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 123,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "",
              "pathname": "",
              "search": "",
              "state": "",
            },
            "push": [Function],
            "replace": [Function],
          }
    }
    initialValues={
        Object {
            "profileDescription": "",
            "profileTitle": "",
          }
    }
    intl={
        Object {
            "defaultFormats": "",
            "defaultLocale": "",
            "formatDate": [Function],
            "formatHTMLMessage": [Function],
            "formatMessage": [Function],
            "formatNumber": [Function],
            "formatPlural": [Function],
            "formatRelative": [Function],
            "formatTime": [Function],
            "formats": "",
            "locale": "",
            "messages": Object {},
            "now": [Function],
          }
    }
    isSubmitting={false}
    isValid={false}
    isValidating={false}
    location={
        Object {
            "hash": "",
            "key": "",
            "pathname": "",
            "search": "",
            "state": "",
          }
    }
    match={
        Object {
            "isExact": true,
            "params": Object {},
            "path": "",
            "url": "",
          }
    }
    onSubmit={[Function]}
    registerField={[Function]}
    resetForm={[Function]}
    setError={[Function]}
    setErrors={[Function]}
    setFieldError={[Function]}
    setFieldTouched={[Function]}
    setFieldValue={[Function]}
    setFormikState={[Function]}
    setStatus={[Function]}
    setSubmitting={[Function]}
    setTouched={[Function]}
    setValues={[Function]}
    submitCount={0}
    submitForm={[Function]}
    touched={Object {}}
    unregisterField={[Function]}
    user={
        Object {
            "_id": "id",
            "address": "address",
            "availability": undefined,
            "avatar": "avatar",
            "car": false,
            "children": Array [
              Object {
                "dob": 122322,
                "gender": "boy",
                "id": "childId",
                "information": "childInformation",
                "name": "childName",
                "school": "school",
              },
            ],
            "city": "city",
            "drivingLicense": false,
            "email": "email",
            "firstName": "firstName",
            "lastName": "lastName",
            "lat": 1234,
            "lng": 123445,
            "nonSmoker": false,
            "photoURL": "photU\`rl",
            "postCode": "postCode",
            "profileDescription": "profileDescription",
            "profileTitle": "profileTitle",
            "unavailability": undefined,
          }
    }
    validateField={[Function]}
    validateForm={[Function]}
    values={
        Object {
            "profileDescription": "profileDescriptionTest",
            "profileTitle": "profileTitleTest",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "enableReinitialize": false,
      "initialValues": Object {
        "profileDescription": "profileDescription",
        "profileTitle": "profileTitle",
      },
      "isInitialValid": false,
      "onSubmit": [Function],
      "render": [Function],
      "validateOnBlur": true,
      "validateOnChange": true,
      "validationSchema": ObjectSchema {
        "_blacklist": RefSet {
          "list": Set {},
          "refs": Map {},
        },
        "_conditions": Array [],
        "_defaultDefault": [Function],
        "_deps": Array [],
        "_excludedEdges": Array [],
        "_exclusive": Object {},
        "_mutate": false,
        "_nodes": Array [
          "profileDescription",
          "profileTitle",
        ],
        "_options": Object {
          "abortEarly": true,
          "recursive": true,
        },
        "_type": "object",
        "_typeError": [Function],
        "_whitelist": RefSet {
          "list": Set {},
          "refs": Map {},
        },
        "fields": Object {
          "profileDescription": StringSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "required": false,
            },
            "_mutate": false,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "string",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
          "profileTitle": StringSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "required": false,
            },
            "_mutate": false,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "string",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
        },
        "tests": Array [],
        "transforms": Array [
          [Function],
        ],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "enableReinitialize": false,
        "initialValues": Object {
          "profileDescription": "profileDescription",
          "profileTitle": "profileTitle",
        },
        "isInitialValid": false,
        "onSubmit": [Function],
        "render": [Function],
        "validateOnBlur": true,
        "validateOnChange": true,
        "validationSchema": ObjectSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_defaultDefault": [Function],
          "_deps": Array [],
          "_excludedEdges": Array [],
          "_exclusive": Object {},
          "_mutate": false,
          "_nodes": Array [
            "profileDescription",
            "profileTitle",
          ],
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "object",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "fields": Object {
            "profileDescription": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": false,
              },
              "_mutate": false,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
            },
            "profileTitle": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": false,
              },
              "_mutate": false,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
            },
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
